### ISSUE TITLE
How to create a simple Intraprocedural Checker with Infer.AI

### ISSUE THREAD

**Issue by Rui1995** [facebook/infer#1225]
Hi, 
I'm starting to learn how to use the Infer.AI framework, and am trying to create a simple intraprocedural checker.
I was following the instructions from https://fbinfer.com/docs/adding-checkers.html , but they seem to be outdated (for example it's refered to "PatternMatch.java_proc_name_with_class_method" that doesn't exist anymore).

Can someone explain me what are the basic "ingredients" to making an intraprocedural checker with Infer.AI?

Sorry if this is a very basic question, but I can't seem to find an updated answer online for it.

**Comment by dulmarod** [comment]
Hi, you are right that the documentation is a bit outdated, we are planning to update it soon. 

I can point you to the main blocks of creating an intra-procedural checker, based on the checker SelfInBlock that I wrote recently. See infer/src/checkers/SelfInBlock.ml. 

```
	
module Domain = AbstractDomain.Map (Ident) (DomainData)
(* An example domain, where DomainData is a module that needs to implement the functions 

pp
join
leq
widen 

In my case, widen is just 

  let widen ~prev ~next ~num_iters:_ = join prev next

Other examples of domains are any module where you implement those functions, or some other modules in AbstractDomain, like FiniteSet, etc.
*)

module TransferFunctions = struct
  module Domain = Domain
  module CFG = ProcCfg.Normal

  type extras = unit

  let pp_session_name _node fmt = F.pp_print_string fmt "SelfCapturedInBlock"

  let exec_instr (astate : Domain.t) {ProcData.summary} _cfg_node (instr : Sil.instr) = ...
(* Here you pattern match over the Sil instructions and update the domain *)

module Analyzer = AbstractInterpreter.MakeWTO (TransferFunctions)

let checker {Callbacks.exe_env; summary} =
  let initial = (* initial state of your domain, say a map *)
  let procname = Summary.get_proc_name summary in
  let tenv = Exe_env.get_tenv exe_env procname in
  let proc_data = ProcData.make summary tenv () in
  match Analyzer.compute_post proc_data ~initial with
    | Some domain ->
        report_issues summary domain
    | None ->
        () ;
  summary
```

As for `report_issues` , the main parts for logging issues are 

```
    let message = "This is a bug" in
    Reporting.log_error summary ~loc IssueType.captured_strong_self
      message ;
```

You need to pass the location of the issue, and also create a new issue type in `IssueType` module. 

I realise this is not trivial at all, if you still want to proceed from this introduction, feel free to get in touch with more questions.

**Comment by Rui1995** [comment]
Thank you for the information.
I'm going to follow this introduction and, if you don't mind, contact you if I have any doubts.

Best Regards.

**Comment by Rui1995** [comment]
Hi,
I've been trying to follow your instructions, but I've come across an error I've been unable to fix.

I've created a simple checker, and the code is the following: https://gist.github.com/Rui1995/ba9705060fc156b41a7c4ef32cd94a91  .
The checker is already registered, and the mli file is the exact same as SelfInBlock.mli .
The error I've come across is: https://gist.github.com/Rui1995/7ddd37b92d2fa2ab977a1622b8c7f803   .
It seems Infer doesn't acknowledge that Domain.t is an integer.

Again, sorry if this is a basic error, but I've spent an entire day trying to fix it with no success.

Best Regards.

**Comment by jvillard** [comment]
> `module Domain = AbstractDomain.Map (Ident) (DomainData)`

Your domain is a map from `Ident` to `int`, not an `int`. Try replacing `let initial = 0` by `let initial = Domain.empty`?

**Comment by Rui1995** [comment]
Thank you,
You were absolutely right. I hadn't figured out exactly what that call to AbstractDomain did.

Best Regards.

**Comment by jiseongg** [comment]
Hi @Rui1995 ,

I've been struggling with same problem. So I'm writing some simple questions here.

My abstract domain is also map like yours
> module Domain = AbstractDomain.Map (Ident) (DomainData)

And, I wanna add some `(Ident, DomainData)` to `astate` in `exec_instr`. However, it seems that `Domain.add` is not visible from `TransferFunctions`. What should I do for it? It seems that not only `add`, but also all other utils for `AbstractDomain.Map` are not accessible. (Actually it's quite difficult to trace `AbstractDomain.Map`'s core operators due to long `include` chain starting from `AbstractDomain.ml`. Thus, I'm not sure I'm in right approach)

Best regards

**Comment by jvillard** [comment]
@jiseongg if you can share the code maybe we could tell what's wrong.

> Actually it's quite difficult to trace AbstractDomain.Map's core operators due to long include chain starting from AbstractDomain.ml

You should be able to jump to definition using [merlin](https://github.com/ocaml/merlin/), which supports a range of editors.

**Comment by jiseongg** [comment]
@jvillard I've just uploaded code snippets in the gist, little bit wondering that some minor syntax errors exist. (I've selectively copy-and-pasted my original code)

https://gist.github.com/jiseongg/b82dec1a63f20854855604c05ab68b69

Short description for code snippets
- Abstract state is memory mapping from `Ident` to `DomainData`
- `DomainData`  is either `Nullable` or `NonNull`
- The most curious part is the line 21-22 in ToyChecker.ml, where I'm adding `(Ident, Nullable)` or `(Ident, NonNull)` to `astate`, depending on whether function call is passing pointers as arguments.

I thought that I can use `Domain.add` in `ToyChecker.ml` but I cannot. I thought the phrase `include AbstractDomain.Map (Ident) (DomainData)` seemed to expose all of functions inside `AbstractDomain.Map`. When I jumped a lot with merlin, it seemed that all those functions are from `Caml.Map.S`

**Comment by jvillard** [comment]
@jiseongg one thing that jumps out at me:  ToyCheckerDomain.mli says `include AbstractDomain.S`. This doesn't expose the types of keys and values of the map, you need to add type constraints, or change it to something like `include module type of struct AbstractDomain.Map (Ident) (DomainData) end`. You could try deleting ToyCheckerDomain.mli altogether to begin with, this will un-hide all the details of ToyCheckerDomain.ml.

**Comment by jiseongg** [comment]
@jvillard Thank you, it was a great hint!! :) The problem I faced has just been resolved.

I think I've misunderstood OCaml module system and how Infer is using it. I fixed the codes and uploaded them to my gist([link](https://gist.github.com/jiseongg/b82dec1a63f20854855604c05ab68b69)). I just attaches the most important parts of my implementation below, to possibly help someone designing analyzer with Infer's internal OCaml modules API.

```ocaml
(* ToyCheckerDomain.mli *)
module DomainData : AbstractDomain.S

include module type of AbstractDomain.Map (Ident) (DomainData)


(* ToyCheckerDomain.ml *)
include AbstractDomain.Map (Ident) (DomainData)
```

And I can now access core functions of `AbstractDomain.Map (Ident) (DomainData)` in my `TransferFunctions`, defined in `ToyChecker.ml`.

<img width="1101" alt="image" src="https://user-images.githubusercontent.com/47690444/113383807-a919a980-93bf-11eb-8f06-07375e8507e8.png">
